File: packages/formspec-layout/src/tokens.ts
/**
 * Resolve a `$token.xxx` reference against component and theme token maps.
 *
 * Component tokens take precedence over theme tokens. Values that are not
 * `$token.` prefixed strings pass through unchanged. Logs a warning when
 * a token reference cannot be resolved in either map.
 *
 * @param val             - The value to resolve. Only strings starting with `$token.` are looked up.
 * @param componentTokens - Token map from the component document (higher priority).
 * @param themeTokens     - Token map from the theme document (lower priority).
 * @returns The resolved token value, or the original value if it is not a token reference.
 */
export function resolveToken(
    val: any,
    componentTokens: Record<string, string | number> | undefined,
    themeTokens: Record<string, string | number> | undefined,
): any {
    if (typeof val === 'string' && val.startsWith('$token.')) {
        const tokenKey = val.substring(7);
        if (componentTokens && componentTokens[tokenKey] !== undefined) {
            return componentTokens[tokenKey];
        }
        if (themeTokens && themeTokens[tokenKey] !== undefined) {
            return themeTokens[tokenKey];
        }
        console.warn(`Unresolved token reference: ${val}`);
    }
    return val;
}

/**
 * Emit merged theme + component tokens as `--formspec-*` CSS custom properties on `target`.
 * Component tokens override theme tokens (same merge order as `emitTokenProperties` in the web component).
 */
export function emitMergedThemeCssVars(
    target: HTMLElement,
    options: {
        themeTokens?: Record<string, string | number> | null;
        componentTokens?: Record<string, string | number> | null;
    },
): void {
    const merged = {
        ...(options.themeTokens || {}),
        ...(options.componentTokens || {}),
    };
    for (const [key, value] of Object.entries(merged)) {
        target.style.setProperty(`--formspec-${key.replace(/\./g, '-')}`, String(value));
    }
}

File: crates/formspec-theme/src/tokens.rs
//! Token resolution: $token.key references against component > theme > renderer defaults.
//! Non-recursive — detects circular references and returns None.

use serde_json::{Map, Value};

const TOKEN_PREFIX: &str = "$token.";

/// Resolve a `$token.key` reference against component tokens, theme tokens,
/// and renderer defaults. Non-recursive. Returns None if unresolved.
pub fn resolve_token(
    value: &str,
    component_tokens: Option<&Map<String, Value>>,
    theme_tokens: Option<&Map<String, Value>>,
) -> Option<Value> {
    if !value.starts_with(TOKEN_PREFIX) {
        return None;
    }

    let key = &value[TOKEN_PREFIX.len()..];
    if key.is_empty() {
        return None;
    }

    // Component tokens have highest priority
    if let Some(ct) = component_tokens {
        if let Some(v) = ct.get(key) {
            // Non-recursive: if the resolved value is itself a token ref, warn and stop
            if let Some(s) = v.as_str() {
                if s.starts_with(TOKEN_PREFIX) {
                    // Recursive token detected — return None
                    return None;
                }
            }
            return Some(v.clone());
        }
    }

    // Theme tokens
    if let Some(tt) = theme_tokens {
        if let Some(v) = tt.get(key) {
            if let Some(s) = v.as_str() {
                if s.starts_with(TOKEN_PREFIX) {
                    return None;
                }
            }
            return Some(v.clone());
        }
    }

    None
}

/// Resolve all $token references in a style map.
pub fn resolve_style_tokens(
    style: &Map<String, Value>,
    component_tokens: Option<&Map<String, Value>>,
    theme_tokens: Option<&Map<String, Value>>,
) -> Map<String, Value> {
    let mut resolved = Map::new();
    for (key, val) in style {
        if let Some(s) = val.as_str() {
            if let Some(resolved_val) = resolve_token(s, component_tokens, theme_tokens) {
                resolved.insert(key.clone(), resolved_val);
                continue;
            }
        }
        resolved.insert(key.clone(), val.clone());
    }
    resolved
}
